gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -248,24 +252,50 @@ private WriteObjectResponse doResumableUpload() throws IOException {
-        WriteObjectRequest insertRequest;
+        WriteObjectRequest insertRequest = null;
         if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
           insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
           writeOffset += insertRequest.getChecksummedData().getContent().size();
         } else {
-          ByteString data =
-              ByteString.readFrom(
-                  ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE), MAX_BYTES_PER_MESSAGE);
-          dataChunkMap.put(writeOffset, data);
           if (dataChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
-            dataChunkMap.remove(dataChunkMap.firstKey());
+            /*
+             If dataChunkMap is full, get the committedWriteOffset. This will make a API call to the
+             server and add latency in this path/context. The upload is happening asynchronously in
+             a different context. So this API will not add latency to overall upload. It will only
+             throttle the upstream write call, which is fine.
+            */
+            committedWriteOffset = getCommittedWriteSizeWithRetries(uploadId);
+            logger.atFinest().log(
+                "Fetched committedWriteOffset: size:%d, numBuffers:%d, writeOffset:%d, committedWriteOffset:%d",
+                dataChunkMap.size(),
+                channelOptions.getNumberOfBufferedRequests(),
+                this.writeOffset,
+                committedWriteOffset);
+
+            // check and remove chunks from dataChunkMap
+            while (dataChunkMap.size() > 0 && dataChunkMap.firstKey() < committedWriteOffset) {
+              logger.atFinest().log(
+                  "clearing dataChunkMap one buffer at a time, size: %d, firstKey:%d, committedwriteOffset:%d",
+                  dataChunkMap.size(), dataChunkMap.firstKey(), committedWriteOffset);
+              dataChunkMap.remove(dataChunkMap.firstKey());
+            }
           }
-          insertRequest = buildInsertRequest(writeOffset, data, false);
-          writeOffset += data.size();
-        }
-        requestStreamObserver.onNext(insertRequest);
-        objectFinalized = insertRequest.getFinishWrite();
 
+          // Pick up a chunk to write only if dataChunkMap has space. Else continue after looking
+          // for errors.
+          if (dataChunkMap.size() < channelOptions.getNumberOfBufferedRequests()) {
+            ByteString data =
+                ByteString.readFrom(
+                    ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE), MAX_BYTES_PER_MESSAGE);
+            insertRequest = buildInsertRequest(writeOffset, data, false);
+            dataChunkMap.put(writeOffset, insertRequest);
+            writeOffset += data.size();
+          }
+        }
+        if (insertRequest != null) {
+          requestStreamObserver.onNext(insertRequest);
+          objectFinalized = insertRequest.getFinishWrite();
+        }
```

Comment:
Can this be moved to previous conditional block ?

Reply:
insertRequest is being set in 2 out of 3 conditional blocks above. Code will repeat if I move it up.